testing/python/metafunc.py

```diff
@@ -1784,3 +1832,39 @@ def test_foo(a):
         )
         result = testdir.runpytest()
         result.assert_outcomes(passed=1)
+
+    def test_parametrize_iterator(self, testdir):
+        testdir.makepyfile(
+            """
+            import itertools
+            import pytest
+
+            id_parametrize = pytest.mark.parametrize(
+                ids=("param%d" % i for i in itertools.count()
+            ))
+
+            @id_parametrize('y', ['a', 'b'])
+            def test1(y):
+                pass
+
+            @id_parametrize('y', ['a', 'b'])
+            def test2(y):
+                pass
+
+            @pytest.mark.parametrize("a, b", [(1, 2), (3, 4)], ids=itertools.count(1))
+            def test_converted_to_str(a, b):
+                pass
+        """
+        )
+        result = testdir.runpytest("-vv", "-s")
+        result.stdout.fnmatch_lines(
+            [
+                "test_parametrize_iterator.py::test1[param0] PASSED",
```

Comment:
i would argue that this is wrong,
its misleading with infinite generators, and will break horrible with other generators
additionally this will potentially cause ids to be different based on how many test files are considered for collection

Reply:
@RonnyPfannschmidt 
Thanks for the feedback!

I see that you can shoot yourself in the foot with this somehow, however: collection happens on the whole file, so it's only relevant when using this across files, and then it would be the same when wrapping it in something to get the same outcome.

This PR/feature is about making it easier to use generators/iterators.

As for infinite generators that should not be a problem, or am I missing something?

But it certainly shows that collection is not stable for py35, likely due to some dict not being ordered somewhere there.  IIRC the test using the itertools fixture shows that it might sometimes collect/generate test2 before test1 there.

Reply:
again, its already wrong on a single file and gets even more wrong on multiple files
as  far as i can tell, the only way to support iterators sanely would be to materealize them exactly once, that's not done here

Reply:
> again, its already wrong on a single file

Like I've said already this appears to be the case for py35 only, and also happens with:
```python
import pytest

id = 0


def idfn(val):
    global id
    id += 1
    return id


@pytest.mark.parametrize("a", [1], ids=idfn)
def test1(a):
    pass


@pytest.mark.parametrize("a", [2], ids=idfn)
def test2(a):
    pass
```
i.e. it's due to inconsistent collection there.
Or do you mean something else?

I can see that it could also be wrong when having some conditional code (i.e. test functions only defined on some conditions etc (where the decorator is used less often)) though.
But similar to using an iterator across multiple files this is something that happens by (abusing) the `id` as a function already.

> the only way to support iterators sanely would be to materealize them exactly once, that's not done here

I do not think that's possible really, is it?

Should we have something special to support the case of "just provide numeric ids, with a possible prefix"?

The next planned step here would have been to have a idsetfn function (or support it for `ids`), where this could be achieved through then maybe.
Should I go with `ids/idsetfn=lambda index: "param" + str(index)` then only?

Reply:
You just use a broken example to justify a broken feature

It would be more apparent if the same mark was used,  and even more apparent if the iterator version was size limited to the length of the parameter set 

This is a no go

Reply:
I'm not sure I follow, but is this example less/not broken then?  (it also shows inconsistent results with py35):

```python
import pytest

id = 0


def idfn(val):
    global id
    id += 1
    return id


id_param = pytest.mark.parametrize("a", [1], ids=idfn)


@id_param
def test1(a):
    pass


@id_param
def test2(a):
    pass
```

Anyway, I've asked already which direction I should follow here, i.e. by not supporting iterators in the first place, but only a new / overloaded idfnset / ids function, which gets the index etc.

Reply:
the example is broken in that it turns ids (which should related to parameters) into something thats related to a uncontrolled global variable 

what happens with the generator support you propose here is that this uncontrolled global variable with effects at a distance gets hidden in a mark 

as far as im concerned the only way to support iterators is bounded materialization that happens once, aka impossible with the current mark backend

Reply:
Ok.

I see three options then:

a) only allow generators in the top-most mark's args/kwargs (when generating the final mark), i.e. it would not allow them in the example above, raising an error.
b) store the generated results on the mark it was given for.  This would require some code changes (currently mark's args/kwargs are passed around, not the Mark itself).  I will push some WIP for this.
c) do not support this (as of now), closing https://github.com/pytest-dev/pytest/issues/759 as wont-fix.

Reply:
D) defer this for after object based markers

Reply:
And how would it be done then?  From https://github.com/pytest-dev/pytest/issues/5424 it seems like this could be achieved for ParametrizedMark already (like mentioned below, where for `name == "parametrize"` it would instantiate that).

Reply:
its a potentially acceptable intermediate hack i would say